File: 03Better/Source.cpp
#include <iostream>
#include <assert.h>

class Welford {
public:
	void Add(double x) {
		count++;
		const double last_mean = mean;
		mean = mean + (x - mean) / count;
		sum = sum + (x - last_mean) * (x - mean);
	}
	int Count() const {
		return count;
	}
	double Mean() const {
		return mean;
	}
	double Var() const {
		return ((count > 1) ? sum / (count - 1) : 0.0);
	}
	double Std() const {
		return sqrt(Var());
	}
private:
	int count = 0;
	double mean = 0, sum = 0;
};

void main() {
	Welford object;
	object.Add(3);
	assert(object.Mean() == 3); // =AVERAGE(3)
	assert(object.Var() == 0); // =VAR.S(3)
	object.Add(7);
	assert(object.Mean() == 5); // =AVERAGE(3,7)
	assert(object.Var() == 8); // =VAR.S(3,7)
	object.Add(11);
	assert(object.Mean() == 7); // =AVERAGE(3,7,11)
	assert(object.Var() == 16); // =VAR.S(3,7,11)
	std::cout << "ok";
	std::cin.get();
}

File: 03Better/source.py
class Welford(object):
    def __init__(self):
        self.count = 0
        self.mean = 0
        self.sum = 0
    def Add(self, x):
        try:
            float(x)
            self.count += 1
            last_mean = self.mean
            self.mean = self.mean + (x - self.mean) * 1.0 / self.count
            self.sum = self.sum + (x - last_mean) * (x - self.mean)
        except ValueError:
            pass
    @property
    def Count(self):
        return self.count
    @property
    def Mean(self):
        return self.mean
    @property
    def Var(self):
        return 0 if self.count<=1 else self.sum/(self.count-1)
        
def main():
    assert True
    object = Welford()
    object.Add(3)
    assert object.Mean == 3
    assert object.Var == 0
    object.Add(7)
    assert object.Mean == 5
    assert object.Var == 8
    object.Add(11)
    assert object.Mean == 7
    assert object.Var == 16
    print('ok')
        
if __name__=='__main__':
    main()